### PR TITLE
Add tfsec and run prior to executing terraform

### PR DESCRIPTION
**Pull request by JonZeolla**
# Contributor Comments
This adds `tfsec` to the image and runs (interactively and non-interactively) prior to executing `terraform` commands.

## Testing
1. Build the new image via `make build`
1. Test the container in a few ways:
    1. Ensure the updates still work via `make update`
    1. 
    1. 
    1. 
    1. 

## Pull Request Checklist
Thank you for submitting a contribution to `easy_infra`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:
 - [X] Ensure that `make build` completes successfully
 - [X] Rebase your branch against the latest commit of the target branch, which likely should be `master`
 - [X] If there is an issue associated with your Pull Request, align your PR title with [this documentation](https://help.github.com/en/articles/closing-issues-using-keywords)

### CODE REVIEW

**Comment by JonZeolla** [comment]
@pb-seiso Ready for review

File: Makefile [hunk 1]
```diff
@@ -19,21 +22,51 @@ $(error COMMIT_HASH was not properly set)
...
+.PHONY: update-functions
+update-functions:
+	@echo "Updating the functions script..."
+	@echo "# DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN" > functions
+	@cat functions_static >> functions
+	@chmod 0755 functions
+	@# This method works as long as FUNCTION_GENERATOR is treated like a list of 4-tuples
+	@# It was chosen because Make and bash don't natively support multidimensional arrays
+	@i=0; for element in $(FUNCTION_GENERATOR); do \
+		if (( i % 4 == 0 )) ; then \
```
Comment (pb-seiso) [inline Makefile:51]: I'm having trouble understanding this check. Why isn't the warning line just printed once before the `for` loop?

File: functions [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  security_tool_argument="--skip-${security_tool}"
+  security_tool_env_var="SKIP_${security_tool_upper}"
+
+  if [[ "${1}" == "${security_tool_argument}" ]]; then
+    _feedback WARNING "Skipping ${security_tool} due to ${security_tool_argument} argument"
+    shift
+  elif [[ "${!security_tool_env_var:-false}" == "true" ]]; then
+    _feedback WARNING "Skipping ${security_tool} due to the ${security_tool_env_var} environment variable value"
+  elif [[ -r "/${security_tool}_complete" ]]; then
+    _feedback INFO "Skipping ${security_tool} because it was already run on $(date -d @$(cat /${security_tool}_complete))"
```
Comment (pb-seiso) [inline functions:53]: Is there any drawback to running tfsec every time? What if your template changes during the same session and the `tfsec_complete` from the last run is still present? Or if a box is compromised and your Terraform environment variables or .tf are modified?

File: functions [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  if [[ ! "${-}" =~ .*i.* ]]; then
+    date +%s > "/${security_tool}_complete"
+  fi
+
+  return ${exit_code}
+}
+
+# DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
+function terraform() {
+  security_tool="tfsec"
```
Comment (pb-seiso) [inline functions:40]: Breaking out each of these into variables makes the script less readable to me, and it's hard to follow what exactly is being called. The function is called `terraform()`, so I don't see any issue putting `tfsec`  and its arguments in-line.

File: functions_template [hunk 4]
```diff
@@ -0,0 +1,28 @@
+function %TODO%() {
```
Comment (pb-seiso) [inline functions_template:1]: Is this function valuable to leave in? Any future tools probably aren't going to be generalized with the same calling conventions as Terraform.

Comment (JonZeolla) [reply functions_template:1]: The function name would map 1:1 with the name of the tool in use.  The following 3 `%TODO%`s allow you to specify the security checking tool that maps to the tool, the arguments to that tool, and then a human-readable explanation of what the security tool is attempting to do, respectively.  For instance, we may want to scan `ansible` code in the future, and I could envision this being useful with little to no modification in that case

File: functions (hunk 3 above)
Comment (JonZeolla) [reply functions:40]: Do you mean move that logic to the `make generate` so the resulting function is more readable?  That makes sense to me.

File: functions (hunk 2 above)
Comment (JonZeolla) [reply functions:53]: Just the overhead if there are many commands chained together.  If you're running interactively `tfsec_complete` is never created, it only applies non-interactively.  The assumption is that you won't be running something like `docker run easy_infra "terraform thing; sed -i 's/a/b/' file; terraform thing2"`

File: Makefile (hunk 1 above)
Comment (JonZeolla) [reply Makefile:51]: I'm leaning towards over-communicating.  This approach makes sure the notice gets added before the static template and then directly before each generated function.

**Comment by JonZeolla** [comment]
Agree on bash vs python in general - bash is not a great general purpose programming language.  My decision for sticking with bash was that it is native and doesn't depend on python, so could be used if we decide python is not needed for this container, or some other container which reuses the logic developed here.  `ubuntu:20.04` does not come with `python` or `python3`, if it did I would prefer to go that route.  For now I'd like to stick with `bash` but improve readability.  Let me take a stab at that

**Comment by JonZeolla** [comment]
@pb-seiso back to you

**Comment by pb-seiso** [comment]
@JonZeolla The changes look good, especially around commenting and simplifying some of the bash-fu. Approved and good to merge.

**Comment by JonZeolla** [comment]
@pb-seiso Identified a bug with my prior push, updated and retested.  Merging soon

**Comment by pb-seiso** [comment]
@JonZeolla Approved
